src/reconcile.py

```diff
@@ -0,0 +1,274 @@
+        level=path_info.level,
+        path=path_info.table_path,
+        navlink_title=path_info.navlink_title,
+        content=path_info.local_path.read_text() if path_info.local_path.is_file() else None,
+    )
+
+
+def _local_and_server(
+    path_info: types_.PathInfo, table_row: types_.TableRow, discourse: Discourse
+) -> (types_.UpdateAction | types_.NoopAction | types_.CreateAction | types_.DeleteAction):
+    """Return an update or noop action depending on if the content or navlink title has changed.
+
+    Args:
+        path_info: Information about the local documentation file.
+        table_row: A row from the navigation table.
+        discourse: A client to the documentation server.
+
+    Returns:
+        A page update or noop action.
+
+    Raises:
+        ReconcilliationError: if the table path or level do not match for the path info and table
+            row.
+    """
+    if path_info.level != table_row.level:
+        raise exceptions.ReconcilliationError(
+            f"internal error, level mismatch, {path_info=!r}, {table_row=!r}"
+        )
+    if path_info.table_path != table_row.path:
+        raise exceptions.ReconcilliationError(
+            f"internal error, table path mismatch, {path_info=!r}, {table_row=!r}"
+        )
+
+    # Is a directory locally and a grouping on the server
+    if path_info.local_path.is_dir() and table_row.is_group:
+        if table_row.navlink.title == path_info.navlink_title:
+            return types_.NoopAction(
+                action=types_.Action.NOOP,
+                level=path_info.level,
+                path=path_info.table_path,
+                navlink=table_row.navlink,
+                content=None,
+            )
+        return types_.UpdateAction(
+            action=types_.Action.UPDATE,
+            level=path_info.level,
+            path=path_info.table_path,
+            navlink_change=types_.NavlinkChange(
+                old=table_row.navlink,
+                new=types_.Navlink(title=path_info.navlink_title, link=table_row.navlink.link),
+            ),
+            content_change=types_.ContentChange(old=None, new=None),
+        )
+
+    # Is a directory locally and a page on the server
+    if path_info.local_path.is_dir():
+        # This is an edge case that can't actually occur because table_row.is_group is based on
+        # whether the navlink link is None so this case would have been caught in the local
+        # directory and server group case
+        if table_row.navlink.link is None:  # pragma: no cover
+            raise exceptions.ReconcilliationError(
+                f"internal error, expecting link on table row, {path_info=!r}, {table_row=!r}"
+            )
+        return types_.DeleteAction(
+            action=types_.Action.DELETE,
+            level=path_info.level,
+            path=path_info.table_path,
+            navlink=table_row.navlink,
+            content=discourse.retrieve_topic(url=table_row.navlink.link),
```

Comment:
The reasoning for having it is more for usability. For example, the action has a draft mode (in the next PR) that logs all the actions being taken. This is helpful for enabling during a PR for a documentation change or on publishing to edge to make sure all the changes that are being made are as expected. In that case, I thought it would be useful to let the user know the content of the page being deleted. Additionally, if the page is actually deleted  and it was a mistake by the user, they still have the content in the logs so they can recover it. Is that a worthwhile reason to track the content being deleted do you think?

Reply:
Ok, so that's more for "auditing" purpose thant functional purpose. But this is another call to discourse API, so it'll mainly depend on how you the "rate-limit" issue can be handled.
If there's a required compromise I propose to get rid of this one as part of it